File: api/watchlist.py
from http.server import BaseHTTPRequestHandler
import os
import json
import sqlite3
import psycopg2
from urllib.parse import urlparse

# Reuse logic? ideally import, but for serverless single file is safer or use shared module
DATABASE_URL = os.environ.get('POSTGRES_URL')

def get_db_connection():
    if DATABASE_URL:
        return psycopg2.connect(DATABASE_URL)
    return sqlite3.connect('waitlist.db')

def parse_inline_url(url):
    try:
        # https://inline.app/booking/{company_id}/{branch_id}
        parts = url.split('/booking/')
        if len(parts) > 1:
            ids = parts[1].split('/')
            if len(ids) >= 2:
                cid = ids[0]
                bid = ids[1].split('?')[0]
                return cid, bid
    except:
        pass
    return None, None

class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        try:
            content_length = int(self.headers['Content-Length'])
            post_data = self.rfile.read(content_length)
            data = json.loads(post_data.decode('utf-8'))
            
            user_id = data.get('userId')
            booking_url = data.get('bookingUrl')
            target_date = data.get('targetDate')
            party_size = data.get('partySize')
            
            if not user_id or not booking_url:
                self.send_response(400)
                self.end_headers()
                self.wfile.write(b"Missing required fields")
                return

            cid, bid = parse_inline_url(booking_url)
            if not cid:
                self.send_response(400)
                self.end_headers()
                self.wfile.write(b"Invalid Booking URL")
                return

            conn = get_db_connection()
            cursor = conn.cursor()
            
            # 1. Ensure Restaurant Exists
            # Check by branch_id
            query_check = "SELECT id FROM restaurants WHERE branch_id = %s" if DATABASE_URL else "SELECT id FROM restaurants WHERE branch_id = ?"
            cursor.execute(query_check, (bid,))
            row = cursor.fetchone()
            
            if row:
                restaurant_id = row[0]
            else:
                # Insert new
                query_ins = "INSERT INTO restaurants (company_id, branch_id, name, booking_url) VALUES (%s, %s, %s, %s) RETURNING id" if DATABASE_URL else "INSERT INTO restaurants (company_id, branch_id, name, booking_url) VALUES (?, ?, ?, ?)"
                
                if DATABASE_URL:
                     cursor.execute(query_ins, (cid, bid, "Unknown", booking_url))
                     restaurant_id = cursor.fetchone()[0]
                else:
                     cursor.execute(query_ins, (cid, bid, "Unknown", booking_url))
                     restaurant_id = cursor.lastrowid
            
            # 2. Insert Task
            query_task = "INSERT INTO tasks (user_id, restaurant_id, target_date, party_size, status) VALUES (%s, %s, %s, %s, 'PENDING')" if DATABASE_URL else "INSERT INTO tasks (user_id, restaurant_id, target_date, party_size, status) VALUES (?, ?, ?, ?, 'PENDING')"
            cursor.execute(query_task, (user_id, restaurant_id, target_date, party_size))
            
            conn.commit()
            cursor.close()
            conn.close()
            
            self.send_response(200)
            self.end_headers()
            self.wfile.write(json.dumps({"success": True}).encode())
            
        except Exception as e:
            print(f"Error: {e}")
            self.send_response(500)
            self.end_headers()
            self.wfile.write(str(e).encode())

File: api/watchlist.ts
import { sql } from '@vercel/postgres';
import { VercelRequest, VercelResponse } from '@vercel/node';
import * as line from '@line/bot-sdk';

function parseInlineUrl(url: string) {
    try {
        const parts = url.split('/booking/');
        if (parts.length > 1) {
            const ids = parts[1].split('/');
            if (ids.length >= 2) {
                const cid = ids[0];
                const bid = ids[1].split('?')[0];
                return { cid, bid };
            }
        }
    } catch (e) {
        return null;
    }
    return null;
}

export default async function handler(req: VercelRequest, res: VercelResponse) {
    if (req.method !== 'POST') {
        return res.status(405).json({ error: 'Method Not Allowed' });
    }

    try {
        const { userId, bookingUrl, targetDate, partySize } = req.body;

        if (!userId || !bookingUrl) {
            return res.status(400).json({ error: 'Missing required fields' });
        }

        const parsed = parseInlineUrl(bookingUrl);
        if (!parsed) {
            return res.status(400).json({ error: 'Invalid Booking URL' });
        }
        const { cid, bid } = parsed;

        // 1. Ensure Restaurant Exists
        // Check by branch_id
        let restaurantId;
        const checkRes = await sql`SELECT id FROM restaurants WHERE branch_id = ${bid}`;

        if (checkRes.rows.length > 0) {
            restaurantId = checkRes.rows[0].id;
        } else {
            // Insert new
            const insertRes = await sql`
                INSERT INTO restaurants (company_id, branch_id, name, booking_url) 
                VALUES (${cid}, ${bid}, 'Unknown', ${bookingUrl})
                RETURNING id;
            `;
            restaurantId = insertRes.rows[0].id;
        }

        // 2. Insert Task
        await sql`
            INSERT INTO tasks (user_id, restaurant_id, target_date, party_size, status) 
            VALUES (${userId}, ${restaurantId}, ${targetDate}, ${partySize}, 'PENDING');
        `;

        // 3. Send LINE Confirmation
        const lineConfig = {
            channelAccessToken: process.env.LINE_CHANNEL_ACCESS_TOKEN || '',
            channelSecret: process.env.LINE_CHANNEL_SECRET || '',
        };
        const lineClient = new line.Client(lineConfig);

        try {
            await lineClient.pushMessage(userId, {
                type: 'text',
                text: `✅ 預約監控已建立！\n\n日期時段：${targetDate}\n人數：${partySize} 位\n\n系統正在為您監控空位，一旦有釋出將會立即發送通知給您！🚀`,
            });
        } catch (err) {
            console.error('Failed to send LINE confirmation:', err);
        }

        res.status(200).json({ success: true });

    } catch (error) {
        console.error(error);
        res.status(500).json({ error: String(error) });
    }
}
